firedrake/assemble.py

```diff
@@ -90,6 +92,11 @@ def assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
     if len(kwargs) > 0:
         raise TypeError("Unknown keyword arguments '%s'" % ', '.join(kwargs.keys()))
 
+    if get_config()['options']['complex']:
+        if form_compiler_parameters is None:
+            form_compiler_parameters = {}
+        form_compiler_parameters['scalar_type'] = 'double complex'
```

Comment:
This kind of anti-pattern appears *everywhere*.  There should be no need to make these "runtime" decisions at all.  This suggests that almost certainly the scalar type selection in the form compiler does not work right.

I note in passing that if you:

```python
from pyop2.datatypes import IntType, RealType, ScalarType
```

you get numpy dtypes for what PETSc thinks the int, real, and scalar variables are.